cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,354 @@
+  std::vector<int32_t> offsets(num_cells_local + 1);
+
+  {
+    adios2::Variable<std::int64_t> var_input_global_indices
+        = io->InquireVariable<std::int64_t>("input_global_indices");
+
+    adios2::Variable<T> var_x = io->InquireVariable<T>("x");
+
+    adios2::Variable<std::int64_t> var_topology_array
+        = io->InquireVariable<std::int64_t>("topology_array");
+
+    adios2::Variable<std::int32_t> var_topology_offsets
+        = io->InquireVariable<std::int32_t>("topology_offsets");
+
+    if (var_input_global_indices)
+    {
+      var_input_global_indices.SetSelection(
+          {{local_range[0]}, {num_vertices_local}});
+      engine->Get(var_input_global_indices, input_global_indices.data(),
+                  adios2::Mode::Deferred);
+    }
+
+    if (var_x)
+    {
+      var_x.SetSelection({{local_range[0], 0}, {num_vertices_local, 3}});
+      engine->Get(var_x, x.data(), adios2::Mode::Deferred);
+    }
+
+    if (var_topology_array)
+    {
+      var_topology_array.SetSelection({{cell_range[0] * num_dofs_per_cell},
+                                       {cell_range[1] * num_dofs_per_cell}});
+      engine->Get(var_topology_array, array.data(), adios2::Mode::Deferred);
+    }
+
+    if (var_topology_offsets)
+    {
+      var_topology_offsets.SetSelection({{cell_range[0]}, {cell_range[1] + 1}});
+      engine->Get(var_topology_offsets, offsets.data(), adios2::Mode::Deferred);
+    }
+
+    engine->EndStep();
+
+    std::int32_t cell_offset = offsets[0];
+    for (auto offset = offsets.begin(); offset != offsets.end(); ++offset)
+      *offset -= cell_offset;
+  }
+
+  std::vector<T> x_reduced(num_vertices_local * dim);
+  for (std::uint32_t i = 0; i < num_vertices_local; ++i)
+  {
+    for (std::uint32_t j = 0; j < (std::uint32_t)dim; ++j)
+      x_reduced[i * dim + j] = x[i * 3 + j];
+  }
+
+  fem::CoordinateElement<T> element
+      = fem::CoordinateElement<T>(cell_type, degree, lagrange_variant);
+
+  std::array<std::size_t, 2> xshape = {num_vertices_local, (std::uint32_t)dim};
+  auto part = mesh::create_cell_partitioner(mesh::GhostMode::shared_facet);
+
+  mesh::Mesh<T> mesh = mesh::create_mesh(comm, comm, array, element, comm,
+                                         x_reduced, xshape, part);
+  return mesh;
+}
+
+//-----------------------------------------------------------------------------
+/// @cond
+template dolfinx::mesh::Mesh<float> read_mesh<float>(ADIOS2Wrapper& ADIOS2,
+                                                     MPI_Comm comm);
+
+template dolfinx::mesh::Mesh<double> read_mesh<double>(ADIOS2Wrapper& ADIOS2,
+                                                       MPI_Comm comm);
+
+/// @endcond
+
+//-----------------------------------------------------------------------------
+std::string query_type(ADIOS2Wrapper& ADIOS2)
+{
+  auto io = ADIOS2.io();
+  auto engine = ADIOS2.engine();
+  engine->BeginStep();
+  std::string floating_point = io->VariableType("x");
+  engine->EndStep();
```

Comment:
Could this be refactored to start the step, query the dtype, dispatch, read the mesh, then end the step?